src/hotspot/cpu/x86/stubGenerator_x86_64.cpp

```diff
@@ -5472,206 +5401,277 @@ address generate_avx_ghash_processBlocks() {
+      // Each of these byte sequences are then manipulated into 4 6-bit values ready for encoding.
+      //
+      // If we focus on one set of 3-byte chunks, changing the nomenclature such that A0 => a,
+      // A1 => b, and A2 => c, we shuffle such that each 24-bit chunk contains:
+      //
+      // b3 b2 b1 b0 c5 c4 c3 c2 c1 c0 d5 d4 d3 d2 d1 d0 a5 a4 a3 a2 a1 a0 b5 b4 b3 b2 b1 b0 c5 c4 c3 c2
+      //
+      // W first and off all but bits 4-9 and 16-21 (c5..c0 and a5..a0) and shift them using
+      // a vector multiplication operation (vpmulhuw) which effectively shifts c right by 6 bits
+      // and a right by 10 bits.  We similarly mask bits 10-15 (d5..d0) and 22-27 (b5..b0) and
+      // shift them left by 8 and 4 bits respecively.  This is done using vpmullw.  We end up with
+      // 4 6-bit values, thus splitting the 3 input bytes, ready for encoding:
+      //    0 0 d5..d0 0 0 c5..c0 0 0 b5..b0 0 0 a5..a0
+      //
+      // For translation, we recognize that there are 5 distinct ranges of legal Base64
+      // characters as below:
+      //
+      //   +-------------+-------------+------------+
+      //   | 6-bit value | ASCII range |   offset   |
+      //   +-------------+-------------+------------+
+      //   |    0..25    |    A..Z     |     65     |
+      //   |   26..51    |    a..z     |     71     |
+      //   |   52..61    |    0..9     |     -4     |
+      //   |     62      |   + or -    | -19 or -17 |
+      //   |     63      |   / or _    | -16 or 32  |
+      //   +-------------+-------------+------------+
+      //
+      // We note that vpshufb does a parallel lookup in a destination register using
+      // the lower 4 bits of bytes from a source register.  If we use a saturated subtraction
+      // and subtract 51 from each 6-bit value, bytes from [0,51] saturate to 0, and [52,63]
+      // map to a range of [1,12].  We distinguish the [0,25] and [26,51] ranges by
+      // assigning a value of 13 for all 6-bit values less than 26.  We end up with:
+      //
+      //   +-------------+-------------+------------+
+      //   | 6-bit value |   Reduced   |   offset   |
+      //   +-------------+-------------+------------+
+      //   |    0..25    |     13      |     65     |
+      //   |   26..51    |      0      |     71     |
+      //   |   52..61    |    0..9     |     -4     |
+      //   |     62      |     11      | -19 or -17 |
+      //   |     63      |     12      | -16 or 32  |
+      //   +-------------+-------------+------------+
+      //
+      // We then use a final vpshufb to add the appropriate offset, translating the bytes.
+      // 
+      // Load input bytes - only 28 bytes.  Mask the first load to not load into the full register.
+      __ vpmaskmovd(xmm1, xmm1, Address(source, start_offset, Address::times_1, -4), Assembler::AVX_256bit);
+      __ vpshufb(xmm1, xmm1, xmm9, Assembler::AVX_256bit);
+
+      __ addl(start_offset, 24);
+
+      __ movl(rax, 0x003f03f0);
```

Comment:
The big chunk of code from line 5549 to 5568 needs comments.

Reply:
Done.